File: java/parser/src/main/java/gov/lbl/parser/parser/Arithmetic_expressionVisitor.java
package gov.lbl.parser.parser;

import java.util.List;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Arithmetic_expression;
import gov.lbl.parser.domain.Arithmetic_term;
import gov.lbl.parser.domain.Term;

import static java.util.stream.Collectors.toList;

import java.util.ArrayList;

public class Arithmetic_expressionVisitor extends modelicaBaseVisitor<Arithmetic_expression> {
    @Override
    public Arithmetic_expression visitArithmetic_expression(modelicaParser.Arithmetic_expressionContext ctx) {
        TermVisitor termVisitor = new TermVisitor();
        List<Term> terms = ctx.term() == null ? null : ctx.term()
                                                            .stream()
                                                            .map(t -> t.accept(termVisitor))
                                                            .collect(toList());

        List<String> add_ops = ctx.add_op() == null ? null : ctx.add_op()
                .stream()
                .map(add_op -> add_op.getText())
                .collect(toList());  
        
        List<Arithmetic_term> arithmetic_term_list = new ArrayList<Arithmetic_term>();
        if (terms != null && add_ops != null) {
            if (add_ops.size() == (terms.size() - 1)) {
                arithmetic_term_list.add(new Arithmetic_term(null, terms.get(0)));
            }
            for (int i=0; i<add_ops.size(); i++ ) {
                arithmetic_term_list.add(new Arithmetic_term(add_ops.get(i), terms.get(i)));
            }
        }
        return new Arithmetic_expression(arithmetic_term_list);
    }
}

File: jsParser/parser/Arithmetic_expressionVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Arithmetic_expression = require('../domain/Arithmetic_expression')
const Arithmetic_term = require('../domain/Arithmetic_term')

const TermVisitor = require('./TermVisitor')

class Arithmetic_expressionVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitArithmetic_expression (ctx) {
    const add_ops = []
    const terms = []
    const arithmetic_term_list = []

    if (ctx.add_op()) {
      ctx.add_op().forEach(add => {
        add_ops.push(add.getText())
      })
    }

    if (ctx.term()) {
      const termVisitor = new TermVisitor.TermVisitor()
      ctx.term().forEach(t => {
        terms.push(termVisitor.visitTerm(t))
      })
    }
    if (terms.length > 0) {
      let extra_term = 0
      if (add_ops.length == (terms.length - 1)) {
        arithmetic_term_list.push(new Arithmetic_term.Arithmetic_term(null, terms[0]))
        extra_term = 1
      }
      for (let i = 0; i < add_ops.length; i++) {
        arithmetic_term_list.push(new Arithmetic_term.Arithmetic_term(add_ops[i], terms[extra_term + i]))
      }
    }

    return new Arithmetic_expression.Arithmetic_expression(arithmetic_term_list)
  }
}

Arithmetic_expressionVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitArithmetic_expression = this.visitArithmetic_expression
exports.Arithmetic_expressionVisitor = Arithmetic_expressionVisitor
